lib/launcher/src/Cardano/Launcher.hs

```diff
@@ -89,24 +129,96 @@ instance Buildable Command where
 
 -- | ProcessHasExited is used by a monitoring thread to signal that the process
 -- has exited.
-data ProcessHasExited = ProcessHasExited String ExitCode
-    deriving Show
+data ProcessHasExited
+    = ProcessDidNotStart String IOException
+    | ProcessHasExited String ExitCode
+    deriving (Show, Eq)
 
 instance Exception ProcessHasExited
 
 -- | Run a bunch of command in separate processes. Note that, this operation is
 -- blocking and will throw when one of the given commands terminates. Commands
 -- are therefore expected to be daemon or long-running services.
-launch :: [Command] -> IO ProcessHasExited
-launch cmds = do
-    res <- try $ forConcurrently cmds $ \(Command name args before output) -> do
-        before
-        let process = (proc name args) { std_out = output , std_err = output }
-        withCreateProcess process $ \_ _ _ h -> do
-            code <- waitForProcess h
-            throwIO $ ProcessHasExited name code
-    case res of
-        Left e -> return e
-        Right _ -> error $
+launch :: Trace IO LauncherLog -> [Command] -> IO ProcessHasExited
+launch tr cmds = mapM start cmds >>= waitAny >>= \case
+    (_, Left e) -> return e
+    (_, Right _) -> error $
             "Unreachable. Supervising threads should never finish. " <>
             "They should stay running or throw @ProcessHasExited@."
+  where
+    sleep = forever $ threadDelay maxBound
+    start = async . flip (withBackendProcess tr) sleep
+
+withBackendProcess :: Trace IO LauncherLog -> Command -> IO a -> IO (Either ProcessHasExited a)
```

Comment:
What is the purpose of this? A short doc-comment? 🙏

Reply:
I added a doc string. The important thing is that it is a bracket-style function for handling clean up of resources.